design/Pipelines.md

```diff
@@ -0,0 +1,277 @@
+ - Layout for vertex inputs
+ - Layout for fragment outputs
+ - All the fixed-function state
+
+For simplicity we assume most fixed-function state is created in separate object.
+For example a ```DepthStencilState``` object would be allocated and a pointer to it would be stored in the ```RenderPipelineDescriptor```.
+
+Mismatch:
+ - Metal doesn’t have primitive restart.
+ - Metal doesn’t have a sample mask.
+ - Vulkan can have some state like scissor and viewport set on the pipeline as an optimization on some GPUs.
+
+```cpp
+struct RenderPipelineDescriptor {
+    // Same translation as for compute pipelines
+    ShaderModule vsModule;
+    const char* vsEntryPoint;
+    ShaderModule fsModule;
+    const char* fsEntryPoint;
+    PipelineLayout layout;
+
+    // Pipeline input / outputs
+    InputState* inputState;
+    RenderPass* renderPass;
+    int subpassIndex;
+
+    // Fixed function state
+    DepthStencilState* depthStencil;
+    BlendState* blend[kMaxColorAttachments];
+    PrimitiveTopology topology;
+    // TODO other state: rasterizer state, “multisample state”
+};
+
+
+RenderPipeline CreateRenderPipeline(Device device, const RenderPipelineDescriptor* descriptor);
+```
+
+Translation to the backing APIs would be the following:
+ - **D3D12**: Translates to ```ID3D12::CreateComputePipelineState```
+ - **Metal**: Translates to ```MTLDevice::makeRenderPipelineState```
+ - **Vulkan**: Translates to ```vkCreateGraphicsPipelines```
+
+The translation of individual members of RenderPipelineDescriptor is described below.
+
+### Input state
+
+This describes how the vertex buffers are stepped through (stride, instance vs. vertex, instance divisor), and how the attributes are extracted from the buffers (buffer index, format, offset).
+
+Mismatches:
+ - D3D12 takes the stride along with the vertex buffers in ```ID3D12GraphicsCommandList::IASetVertexBuffers``` whereas Vulkan and Metal take it at pipeline compilation time.
+ - Vulkan doesn’t support a divisor for its step rate.
+
+```cpp
+enum StepRate {
+    StepRateVertex,
+    StepRateInstance,
+};
+
+Enum VertexFormat {
+    // TODO make a list of portable vertex formats
+};
+
+struct InputStateDescriptor {
+    struct {
+        bool enabled;
+        VertexFormat format;
+        int offsetInBuffer;
+        int bufferIndex;
+    } attributes[MAX_ATTRIBUTES];
+
+    struct {
+        StepRate rate;
+        int stride;
+    } buffers[MAX_VERTEX_BUFFERS];
+};
+
+InputState* CreateInputState(Device* device, InputStateDescriptor* descriptor);
+```
+
+Translation to the backing APIs would be the following:
+ - **D3D12**: Translates to a ```D3D12_INPUT_DESC```.
+   Each enabled attribute corresponds to a ```D3D12_INPUT_ELEMENT_DESC``` with ```InputSlot``` being the index of the attribute.
+   Other members of the ```D3D12_INPUT_ELEMENT_DESC``` are translated trivially.
+   The stride is looked up in the pipeline state before calls to ```ID3D12GraphicsCommandList::IASetVertexBuffers```.
```

Comment:
the d3d12 backend may need to defer the setup of vertex buffers until the actual draw call, since any PSO binding between the vertex buffer setup and the draw call would potentially change the strides (assuming we keep the strides in PSO)